### DISCUSSION TITLE
NICE You are the creator of this library and a user of Joplin?

### DISCUSSION THREAD

**Discussion by Wladefant** [personalizedrefrigerator/js-draw#27] [Q&A]
Hey, saw your almost complete (?) Pull request in joplin, wanted to check the library out and voila you are also teh creator of this js library.

My questions: are you a user of Joplin and therefore decided to integrate your library or what did happen. Just you know tell the projects goals and creation story.

Why are there no similar libraries and why is there almost no pen support fro Android except in apps like samsung notes and one note. Is it the difficulty the work flow?

**Comment by personalizedrefrigerator** [comment]
> are you a user of Joplin and therefore decided to integrate your library or what did happen. Just you know tell the projects goals and creation story.

`js-draw` was originally created for [GSoC 2022](https://discourse.joplinapp.org/t/image-editor-project-is-maintaining-a-custom-image-editor-worth-it/26855)  at Joplin.

The project goals include:
- A simple UI
- Supporting a large zoom range (currently supports zoom between $1\cdot 10^{12} \times$ and $2\cdot 10^{-10}\times$)
- Touchscreen and pen support

Things that I would like to implement (but are not currently implemented):
- Ability to link to Joplin notes
    - This would likely involve adding a rich text editor to `js-draw` and providing some way to add plugins to this editor.
- PDF annotation (see https://github.com/personalizedrefrigerator/js-draw/tree/work/pdf-support)
- Math rendering (probably through [MathJAX](https://www.mathjax.org/)) 

> Why are there no similar libraries and why is there almost no pen support fro Android except in apps like samsung notes and one note. Is it the difficulty the work flow?

I'm not sure... For stylus/pen specifically, my guess is the need for specialized hardware or software to send pen events.

**Reply by Wladefant** [reply]
what kind of math rendering are we talking about. You mean you write in latex and then it renders or that the handwriting gets recognized and it automatically renders into the math formula (probably the first one, but the second one is a dream)

**Reply by Wladefant** [reply]
Yes pdf annotion wouldnt be not bad. All the functionality is provided by joplin already (Watching the pdf and saving the changes). 

Soon my university will start, thinking about using joplin AND js draw as my learning tool, so i suppose will soon file real use case bug reports and feature requests

**Reply by personalizedrefrigerator** [reply]
> You mean you write in latex and then it renders or that the handwriting gets recognized and it automatically renders into the math formula

I mean writing in LaTeX or using a tool like [MathQuill](http://mathquill.com/).

I would love to have handwriting recognition! However, it will likely be a huge amount of work.

For future reference, an acronym to Google is "HMER (Handwritten Mathematical Expression Recognition)".

> Yes pdf annotion wouldnt be not bad. All the functionality is provided by joplin already (Watching the pdf and saving the changes).

I expect PDF annotation to be somewhat difficult to implement. [The current work-in-progress implementation](https://github.com/personalizedrefrigerator/js-draw/tree/work/pdf-support) uses [`pdfjs`](https://github.com/mozilla/pdf.js) to load/save PDFs. When I last checked, annotation loading was buggy (e.g. colors not loaded correctly for strokes).

It's possible this will require partially parsing the PDF ([PDF spec](https://github.com/adobe/dc-acrobat-sdk-docs/blob/master/docs/pdfstandards/PDF32000_2008.pdf), [library that does this, but doesn't seem to support Bézier curves in annotations](https://github.com/highkite/pdfAnnotate#createinkannotation)).

For future reference, Bézier curves (in annotations) will probably use the Appearance Streams PDF API (§ 12.5.5 of the PDF spec).

**Reply by Wladefant** [reply]
Wait are you already supporting pdf annotation with pdf.js? Because mozilla pdf viewer is not bad at all. It would be nice if for a start you could implement pdf.js. So for example you can just click on the pdf, but instead of opening an external app, opening with joplin. Do you know why laurent does not use pdf.js as our default pdf viewer?

**Reply by personalizedrefrigerator** [reply]
It's only able to correctly load backgrounds --- the `pdf.js` API I was trying to use for annotations didn't seem to load colors (and, if I remember correctly, there doesn't seem to be a standard way to interpret PDF ink annotations (e.g. as splines or as polylines)). 

This [subproject of Joplin](https://github.com/laurent22/joplin/tree/dev/packages/pdf-viewer) is a pdf-js-based PDF viewer.

[The pdf.js website asks that, before embedding pdf.js in a custom site/app, users "re-skin or build upon it"](https://mozilla.github.io/pdf.js/getting_started/). I think that the re-skinned PDF viewer in the Joplin monorepo was disabled because of rendering bugs and non-working annotation support.
